File: examples/js/ship-nft.ts
/**
 * Example: mint, customise and trade ship NFTs (module: ship_nft).
 *
 *   NEBULA_CONTRACT_ID=C... npx tsx ship-nft.ts
 */
import { Keypair } from "@stellar/stellar-sdk";
import {
  ContractCallError,
  arg,
  config,
  fundWithFriendbot,
  invoke,
  loadPlayer,
  requireContractId,
  view,
} from "./lib/contract";

interface ShipNft {
  id: bigint;
  owner: string;
  ship_type: string;
  hull: number;
  scanner_power: number;
  durability: number;
  max_durability: number;
  metadata: Buffer;
  metadata_uri: Buffer;
}

// ShipError codes (docs/ERROR_CODES.md → ship_nft).
const ShipError: Record<number, string> = {
  1: "ShipAlreadyExists",
  2: "ShipNotFound",
  3: "NotOwner",
  4: "SameOwner",
  5: "BatchLimitExceeded",
  6: "InvalidShipType",
  7: "ReentrancyDetected",
  8: "InvalidMetadataUri",
};

async function main() {
  requireContractId(config.contractId, "NEBULA_CONTRACT_ID");
  const owner = await loadPlayer();

  // Step 1: mint. ship_type is a Symbol: "fighter" | "explorer" | "hauler".
  const ship = await invoke<ShipNft>(owner, "mint_ship", [
    arg.address(owner.publicKey()),
    arg.symbol("explorer"),
    arg.bytes(new Uint8Array()), // on-chain metadata bytes (keep small)
  ]);
  console.log(`Minted ship #${ship.id}: hull=${ship.hull} scanner=${ship.scanner_power}`);

  // Step 2: batch mint (max 3 per tx).
  const fleet = await invoke<ShipNft[]>(owner, "batch_mint_ships", [
    arg.address(owner.publicKey()),
    arg.vec([arg.symbol("fighter"), arg.symbol("hauler")]),
    arg.bytes(new Uint8Array()),
  ]);
  console.log(`Batch minted ${fleet.length} ships`);

  // Step 3: marketplace metadata (ipfs://, https:// or ar://).
  await invoke(owner, "set_metadata", [
    arg.address(owner.publicKey()),
    arg.u64(ship.id),
    arg.bytes("ipfs://bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi"),
  ]);

  // Step 4: list the owner's ships (read-only).
  const ids = await view<bigint[]>(owner, "get_ships_by_owner", [arg.address(owner.publicKey())]);
  console.log(`Owner has ships: ${ids.join(", ")}`);

  // Step 5: transfer to a friend.
  const friend = Keypair.random();
  await fundWithFriendbot(friend.publicKey());
  const moved = await invoke<ShipNft>(owner, "transfer_ship", [
    arg.u64(ship.id),
    arg.address(owner.publicKey()),
    arg.address(friend.publicKey()),
  ]);
  console.log(`Ship #${moved.id} now owned by ${moved.owner}`);

  // Step 6: expected failure. The original owner can't move it again.
  try {
    await invoke(owner, "transfer_ship", [
      arg.u64(ship.id),
      arg.address(owner.publicKey()),
      arg.address(friend.publicKey()),
    ]);
  } catch (e) {
    if (e instanceof ContractCallError && e.code !== null) {
      console.log(`Rejected as expected: ${ShipError[e.code] ?? `#${e.code}`}`);
    } else {
      throw e;
    }
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: examples/ship_nft.rs
//! # Example: Ship NFTs
//!
//! Module: `ship_nft` via `NebulaNomadContract`
//!
//! Shows minting (single and batch), metadata, lookups by owner, transfers,
//! and the `ShipError` codes a dApp should handle.
//!
//! Run:
//! ```text
//! cargo run --example ship_nft
//! ```

use soroban_sdk::{symbol_short, testutils::Address as _, vec, Address, Bytes, Env};
use stellar_nebula_nomad::{NebulaNomadContract, NebulaNomadContractClient, ShipError};

fn main() {
    let env = Env::default();
    env.mock_all_auths();
    let client = NebulaNomadContractClient::new(&env, &env.register(NebulaNomadContract, ()));

    let alice = Address::generate(&env);
    let bob = Address::generate(&env);

    // ── Step 1: Mint a ship ──────────────────────────────────────────────
    // Valid ship types: "fighter", "explorer", "hauler". Each type starts
    // with different hull / scanner stats. `metadata` is free-form bytes
    // stored on-chain (keep it small: storage is paid rent).
    let ship = client.mint_ship(&alice, &symbol_short!("explorer"), &Bytes::new(&env));
    println!(
        "Minted ship #{} ({:?}) hull={} scanner={} durability={}/{}",
        ship.id, ship.ship_type, ship.hull, ship.scanner_power, ship.durability, ship.max_durability
    );

    // ── Step 2: Batch mint (max 3 per transaction) ───────────────────────
    let fleet = client.batch_mint_ships(
        &alice,
        &vec![&env, symbol_short!("fighter"), symbol_short!("hauler")],
        &Bytes::new(&env),
    );
    println!("Batch minted {} ships", fleet.len());

    // More than 3 in one call is rejected with BatchLimitExceeded (#5).
    let too_many = vec![
        &env,
        symbol_short!("fighter"),
        symbol_short!("fighter"),
        symbol_short!("fighter"),
        symbol_short!("fighter"),
    ];
    assert_eq!(
        client.try_batch_mint_ships(&alice, &too_many, &Bytes::new(&env)),
        Err(Ok(ShipError::BatchLimitExceeded))
    );

    // ── Step 3: Attach marketplace metadata ──────────────────────────────
    // URIs must use a marketplace-compatible scheme (ipfs://, https://, ar://).
    let uri = Bytes::from_slice(&env, b"ipfs://bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi");
    client.set_metadata(&alice, &ship.id, &uri);
    println!("Metadata URI set ({} bytes)", client.get_metadata(&ship.id).len());

    let bad_uri = Bytes::from_slice(&env, b"ftp://example.com/ship.json");
    assert_eq!(
        client.try_set_metadata(&alice, &ship.id, &bad_uri),
        Err(Ok(ShipError::InvalidMetadataUri))
    );

    // ── Step 4: Query by owner ───────────────────────────────────────────
    let owned = client.get_ships_by_owner(&alice);
    println!("Alice owns ship ids: {:?}", owned.iter().collect::<std::vec::Vec<_>>());

    // ── Step 5: Transfer ─────────────────────────────────────────────────
    // `from` must sign and must be the current owner.
    let moved = client.transfer_ship(&ship.id, &alice, &bob);
    assert_eq!(moved.owner, bob);
    println!("Ship #{} transferred to Bob", moved.id);

    // Alice no longer owns it: NotOwner (#3).
    assert_eq!(
        client.try_transfer_ship(&ship.id, &alice, &bob),
        Err(Ok(ShipError::NotOwner))
    );
    // Unknown id: ShipNotFound (#2).
    assert_eq!(client.try_get_ship(&9_999u64), Err(Ok(ShipError::ShipNotFound)));

    println!("Bob owns {} ship(s)", client.get_ships_by_owner(&bob).len());
}
